atc/exec/policy_check.go

```diff
@@ -0,0 +1,57 @@
+package exec
+
+import (
+	"github.com/concourse/concourse/atc"
+	"github.com/concourse/concourse/atc/policy"
+)
+
+//go:generate counterfeiter . PolicyChecker
+
+type PolicyChecker interface {
+	Check(teamName, pipelineName, step string, imageSource atc.Source) (bool, error)
+}
+
+type checker struct {
+	policyChecker *policy.Checker
+}
+
+func NewImagePolicyChecker(policyChecker *policy.Checker) PolicyChecker {
+	if policyChecker == nil {
+		return nil
+	}
+	return &checker{policyChecker: policyChecker}
+}
+
+func (c *checker) Check(teamName, pipelineName, step string, imageSource atc.Source) (bool, error) {
+
+	// Actions in skip list will not go through policy check.
+	if c.policyChecker.ShouldSkipAction(policy.ActionUsingImage) {
+		return true, nil
+	}
+
+	imageInfo := map[string]string{
+		"step": step,
+	}
+
+	if repository, ok := imageSource["repository"].(string); ok {
```

Comment:
Maybe we need a way for resource types/prototypes to provide Concourse with some sort of image identifier that can be passed to the policy checker.

In the future maybe this could be returned by the Prototype [`info` call](https://github.com/vito/rfcs/blob/prototypes/037-prototypes/proposal.md#prototype-info)? 🤔 I had a similar thought for a better design for webhooks, where `info` returns some sort of "hook tag" like `github:repo:foo` which can be mapped to incoming web hook payloads in order to know which resources to check.

Reply:
(P.S. I won't block the PR on this if there's really no better way, but I do think we should have a plan for the future and maybe even leave a TODO comment.)

Reply:
Yeah, I think this is a valid comment, we should treat source as a black box. So let me change data sent to policy server to `image_source_type` + `image_source`, where `image_source_type` will be `registry-image`, `docker-image` etc., and `image_source` will be source of the type.